### ISSUE TITLE
Ionic v4.0 android app crash.

### ISSUE THREAD

**Issue by Hitesh61** [bambuser/cordova-plugin-bambuser#5]
When install plugin app run in android device navigate to broadcast page application is crash with following error "java.lang.ClassCastException: android.widget.LinearLayout cannot be cast to android.widget.FrameLayout". 
Android logcat CrashReport.
![CrashReport](https://user-images.githubusercontent.com/36410520/59023263-08aed880-886d-11e9-9e19-e1d6a0dcbe1d.png)

**Comment by tomsun** [comment]
@Hitesh61 Thanks for the report! will investigate.

Which version of Android are you running?

(The output of `ionic info` is always nice to have as well although those values might not have too much of an impact in this case I'm guessing https://ionicframework.com/docs/v3/cli/info/ )

**Comment by Hitesh61** [comment]
Ionic:

   ionic (Ionic CLI)             : 4.12.0
   Ionic Framework           : @ionic/angular 4.1.2
   @angular-devkit/build-angular : 0.12.4
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.2.4
   @ionic/angular-toolkit        : 1.4.0

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : **android 7.1.4**
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 3.1.2, (and 30 other plugins)

System:

   Android SDK Tools : 26.1.1 (E:\Android_Studio)
   NodeJS            : v11.4.0 (C:\Program Files\nodejs\node.exe)
   npm               : 6.4.1
   OS                : Windows 8.1

**Comment by Hitesh61** [comment]
Android app crash issue resolved using change
 FrameLayout layout = (FrameLayout) webView.getView().getParent();
 to
LinearLayout layout = (LinearLayout) webView.getView().getParent();
in CordovaBambuserBroadcaster.java file

But, unable to view broadcast start, stop and switch camera button.

**Comment by tomsun** [comment]
@Hitesh61 thanks for the report!

I added a simple set of cordova-plugin-test-framework tests that run in Travis-CI now, including some runtime checks. They do *not* catch your issue.

(Whatever version of the Android Studio template that https://github.com/apache/cordova-paramedic uses probably has the same `FrameLayout` class as when I use Cordova myself locally)

Also scripted an MVP build test with Ionic's own [Capacitor](https://capacitor.ionicframework.com/), but build-only for now so it doesn't cover this class of error.

When running a Capacitor project locally however, I do get a different but similar failure:
> java.lang.ClassCastException: android.support.design.widget.CoordinatorLayout cannot be cast to android.widget.FrameLayout

I  currently believe casting to ViewGroup is the proper way to go about this, as per https://github.com/bambuser/cordova-plugin-bambuser/commit/4e1791c33170f6812df0c91a2cd4abbcd7c92bc0
This works in the `cordova-paramedic` test harness and in my local Capacitor build as well.

`cordova-plugin-bambuser@0.9.8` is out now and contains that fix. In case you choose to try it out, let me know how it goes.

>  unable to view broadcast start, stop and switch camera button.

Do you have any more details on these issues?

**Comment by Hitesh61** [comment]
I have install cordova-plugin-bambuser@0.9.8 and run project but same issue. camera view start but unable to view broadcast start, stop and switch camera button on top. write now this button's are viewing in behind camera screen.

**Comment by tomsun** [comment]
@Hitesh61 Ok, so the camera view and web view might be stacked in the wrong order in your case.

What is a good starting-point when trying to reproduce? Perhaps:
```
ionic init "Test" --type=angular ...
ionic integrations enable cordova
```

Anything else of importance?

**Comment by Hitesh61** [comment]
ionic start test blank
ionic integrations enable cordova
ionic cordova plugin add cordova-plugin-bambuser@0.9.8
ionic cordova platform add android
ionic cordova run android

**Comment by Hitesh61** [comment]
i think this class are not apply. in demo app class is apply in www/build/main.css, ionic4 build folder is not there inside www folder. 

body.show-viewfinder {
    ion-app,
    ion-content,
    .nav-decor {
      background-color: transparent !important;
    }
  }

**Comment by tomsun** [comment]
@Hitesh61 oh, Capacitor uses a white background for the webview itself!

When I've used Cordova over the years, its Android webview has been transparent by default: hence no explicit request to make it transparent in our plugin, until now in `cordova-plugin-bambuser@0.9.9`, via https://github.com/bambuser/cordova-plugin-bambuser/commit/0f2855f54a51e6d69918f3f96ecf5172678908c1 - does that help in your case?

Perhaps new versions of https://github.com/apache/cordova-android removed the transparency, or perhaps it happens due to the use of https://npmjs.com/package/cordova-plugin-ionic-webview in Ionic 4 projects.

(I didn't manage to verify with Cordova tonight: their generated project doesn't play nice in recent versions of Android Studio, at least for me - something along the lines of https://github.com/apache/cordova-android/issues/718 - and can't do `ionic cordova run/build` locally either ATM for other reasons)

---

In Ionic 4, making the background transparent appears to be a bit easier and more well-supported with the introduction of CSS variables

in `src/global.scss` (for example), try:
```
ion-content{
  --background: transparent;
}
```

This alone seems to do the trick for me.

Btw, I find this snippet helpful for debugging purposes, to make it more obvious whether you've succeeded to make the app transparent, already in the `npm run serve` stage in a web browser:
Makes <html> red (i.e. the bottom-most element within the webview), but also lets some of whatever is behind the webview shine through when you get to that point, thanks to 50% alpha.
```
html {
  background: rgba(255, 0, 0, 0.5); /* just for debugging */
}
```

**Comment by Hitesh61** [comment]
not working. same issue buttons are behind camera view.

**Comment by tomsun** [comment]
I have the Android SDK present for CLI use now to be able to use `cordova build`
instead of Android Studio

---

Steps to verify that the viewfinder is visible:

1. Get an Ionic 4 scaffold
```
ionic start test blank
```

2. Make sample app (semi-)transparent; in `src/global.scss`
```
ion-content {
  --background: transparent;
}

html {
  background: rgba(255, 0, 0, 0.5); /* just for debugging */
}
```

3. Start the viewfinder when app starts; using a snippet from README.md
dumped staight into `src/index.html` for simplicity
```
...
<body>
  <app-root></app-root>
  <script>
    document.addEventListener('deviceready', onDeviceReady, false);
    function onDeviceReady() {
      // Now safe to use device APIs
      if (!window.bambuser) {
        alert('cordova-plugin-bambuser is not installed');
      } else {
        var broadcaster = window.bambuser.broadcaster;
        broadcaster.setApplicationId('YOUR-APPLICATION-ID', function() {
          broadcaster.showViewfinderBehindWebView(function() {
            // maybe also set title etc here before starting
            // maybe move startBroadcast() to a button callback instead of starting right aways
            broadcaster.startBroadcast(function() {
              // update UI to show that we're live!
              alert('We are live!');
            }, function(err) { alert('failed to start broadcast'); });
          }, function(err) { alert('failed to show viewfinder'); });
        }, function(err) { alert('failed to set application id'); });
      }
    }
  </script>
</body>
...
```

4. Add the plugin and build the app:
```
ionic integrations enable cordova
ionic cordova plugin add cordova-plugin-bambuser@0.9.9
ionic cordova platform add android
ionic cordova build android
```

---

I get the expected result, a working viewfinder behind that semitransparent webview, with any of these:
cordova-plugin-bambuser@0.9.7 (does not crash due to the casting in the native Android parts)
cordova-plugin-bambuser@0.9.8 (visible even though we don't make the webview transparent explicitly yet)
cordova-plugin-bambuser@0.9.9

...while using the latest version of the ionic and cordova CLIs:
```
Ionic CLI         : 5.0.3 (/usr/local/lib/node_modules/ionic)
Ionic Framework   : @ionic/angular 4.5.0
Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
Cordova Platforms : android 7.1.4
Cordova Plugins   : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 4.1.0, (and 5 other plugins)
```

if I revert to the versions you specify and start over, i.e
```
npm i --global cordova@8.1.2 ionic@4.12.0
```

...and then start over from step 1 (ionic start etc), I get:
```
ionic (Ionic CLI)     : 4.12.0 (/usr/local/lib/node_modules/ionic)
Ionic Framework       : @ionic/angular 4.5.0
cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
Cordova Platforms     : android 7.1.4
Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 4.1.0, (and 5 other plugins)
```

i.e. newer version of ionic's webview than you have. same version of cordova-android.
No crash with cordova-plugin-bambuser 0.9.7, webview transparency approach still works.

-----------------

Next, downgrading to your version of cordova-plugin-ionic-webview

```
ionic cordova plugin remove cordova-plugin-ionic-webview
ionic cordova plugin add cordova-plugin-ionic-webview@3.1.2
ionic cordova build android
```

Still no crash with cordova-plugin-bambuser 0.9.7, webview transparency approach still works.

**Comment by Hitesh61** [comment]
@tomsun No crash issue in cordova-plugin-bambuser@0.9.8 or cordova-plugin-bambuser@0.9.9, but issue is buttons are behind in camera view. i have use this example https://github.com/bambuser/bambuser-examplebroadcaster-ionic in ionic 4. see below screenshot to identify issue.
![Screenshot_2019-06-21-10-55-07-849_com kasala app](https://user-images.githubusercontent.com/36410520/59899734-6d446880-9413-11e9-8fdf-4a2238407f4c.png)

**Comment by tomsun** [comment]
@Hitesh61 ok: are something funny with that Ionic 3 markup then or is the camera view really on top...? Same result, no red tint, if you're adding this?
```
html {
  background: rgba(255, 0, 0, 0.5); /* just for debugging */
}
```

If no difference, what else could it be...:

We identify the parent of the web view and call `addView()` with index 0: https://github.com/bambuser/cordova-plugin-bambuser/blob/0f2855f54a51e6d69918f3f96ecf5172678908c1/src/android/CordovaBambuserBroadcaster.java#L60
Index 0 for the viewfinder is what should ensure that the webview remains on top. Any possibilities for your FrameLayout to mix things up? Oh, that might be the case actually:

https://developer.android.com/reference/android/widget/FrameLayout

> FrameLayout is designed to block out an area on the screen to display a single item. Generally, FrameLayout should be used to hold a single child view, because it can be difficult to organize child views in a way that's scalable to different screen sizes without the children overlapping each other. 

> You can, however, add multiple children to a FrameLayout and control their position within the FrameLayout by assigning gravity to each child, using the android:layout_gravity attribute. 

So we might need to play with gravity then
https://stackoverflow.com/a/8049860

Would love to know what to do to end up with a FrameLayout in first place though... :-)

**Comment by tomsun** [comment]
@Hitesh61 Wonder if targeting the [parent's parent](https://stackoverflow.com/a/23806038) would make any difference in your case.

Or using various [z-index-affecting](https://stackoverflow.com/questions/6759279/how-to-bring-view-on-front-of-everything) functions on the views involved.

Or moving the webview together with the viewfinder into a wrapper view which we know and control. A bit more intrusive, but might work.

**Comment by Hitesh61** [comment]
Any solution for this issue?

**Comment by tomsun** [comment]
@Hitesh61 my latest post has a few experiments you can try, assuming you are stuck with that FrameLayout.

The preferred solution would be to not to use whatever's causing you to use FrameLayout: use the stock LinearLayout (Cordova) or CoordinatorLayout (Capacitor).
https://www.ibm.com/developerworks/community/blogs/worklight/entry/android_combining_native_and_web_controls_in_cordova_based_applications?lang=en

-------

From your `ionic info` dump:
> cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 3.1.2, (and 30 other plugins)

Mine:
> Cordova Plugins   : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 4.1.0, (and 5 other plugins)

I.e. looks like you have 25 more plugins than I do. 

A. Any of them causing your issues? Do you get the same result with a minimal app as per the steps above?

C. Do you get the same result if you upgrade `ionic` and `cordova` CLI:s to the latest versions?

**Comment by tomsun** [comment]
Closing for now. Feel free to reopen if you have more actionable information @Hitesh61

**Comment by monsieurPaon** [comment]
@Hitesh61 Do you have resolve it ?

**Comment by Hitesh61** [comment]
@monsieurPaon Yes, in my case two plugins 'cordova-admob' and 'cordova-plugin-bambuser' are conflict. Because this two plugin are use same android 'LinearLayout'.

**Comment by monsieurPaon** [comment]
@Hitesh61 thank you for your return. I got a similar issue on ios but not in the same way. I don't have camera preview but i have button and i can record. Unfortunately bambuser team haven't time to help customer.
